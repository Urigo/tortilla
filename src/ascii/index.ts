import * as Fs from 'fs-extra';
import * as Path from 'path';
import {Paths} from '../paths';

/**
  Responsible for printing ascii art files
 */

// Get string for provided ascii art name
function getAsciiArt(artName) {
  const artFile = `${artName}.txt`;
  const artPath = Path.resolve(Paths.tortilla.ascii.views, artFile);

  return Fs.readFileSync(artPath).toString();
}

// Print for provided ascii art
function printAsciiArt(artName) {
  console.log(getAsciiArt(artName));
}

export const Ascii = {
  get: getAsciiArt,
  print: printAsciiArt,
};
