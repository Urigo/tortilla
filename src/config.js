const Handlebars = require('handlebars');
const Paths = require('./paths');
const Renderer = require('./renderer');
const Utils = require('./utils');

let config = {};

if (Utils.exists(Paths.config)) {
  config = require(Paths.config);
}

module.exports = {
  registerCustomTransformations() {
    if (config.render && config.render.transformations) {
      Object.keys(config.render.transformations).forEach((t) => {
        Renderer.registerTransformation(t, 'diffStep', config.render.transformations[t].bind(this, Handlebars));
      });
    }
  },
  getBlacklist() {
    if (config.render && config.render.blacklist) {
      return config.render.blacklist;
    }

    return null;
  },
};
