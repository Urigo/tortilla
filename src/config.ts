import * as Handlebars from 'handlebars';
import { Paths } from './paths';
import { Utils } from './utils';

const Renderer = require('./renderer');

export interface TortillaConfig {
  render?: {
    transformations: any[];
    blacklist: any[];
  }
}

let config: TortillaConfig = {};

if (Utils.exists(Paths.config)) {
  config = require(Paths.config) as TortillaConfig;
}

export function registerCustomTransformations() {
  if (config.render && config.render.transformations) {
    Object.keys(config.render.transformations).forEach((t) => {
      Renderer.registerTransformation(t, 'diffStep', config.render.transformations[t].bind(this, Handlebars));
    });
  }
}

export function getBlacklist() {
  if (config.render && config.render.blacklist) {
    return config.render.blacklist;
  }

  return null;
}

export const Config = config;
