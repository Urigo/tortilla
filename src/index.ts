export * from './config';
export * from './editor';
export * from './essentials';
export * from './git';
export * from './local-storage';
export * from './local-cache';
export * from './manual';
export * from './package';
export * from './paths';
export * from './rebase';
export * from './release';
export * from './step';
export * from './submodule';
export * from './utils';
export * from './translator';
export * from './renderer';
