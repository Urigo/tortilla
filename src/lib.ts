import * as child_process from 'child_process'
import * as stream from 'stream'
import { Dump } from './dump'
import { Git } from './git'

/*
 lib.ts module provides a runtime access to Tortilla's API through a cluster
 to distribute the work to another core and not to block the main thread
**/

// Will call the function through the worker
const callTortillaFn = (fnName, ...args) => {
  return new Promise((resolve, reject) => {
    // Note that we use a child_process here and not a cluster since we don't wanna
    // fork the main module
    const worker = child_process.fork(__filename, [], {
      env: { ...process.env, TORTILLA_WORKER: '1' }
    })

    worker.send({
      op: 'exec',
      fn: fnName,
      args,
    })

    // On result
    worker.on('message', (msg: any) => {
      if (typeof msg !== 'object') { return }
      if (msg.op !== 'result') { return }
      if (msg.fn !== fnName) { return }

      resolve(msg.payload)
    })

    // On error
    worker.on('message', (msg: any) => {
      if (typeof msg !== 'object') { return }
      if (msg.op !== 'error') { return }
      if (msg.fn !== fnName) { return }

      reject(msg.payload)
    })
  })
}

// Helper function that will simulate a registered user in the git.config
const withUser = (fn) => (...args) => {
  let result = ''
  let nameExists = false
  let emailExists = false

  // Upcoming method requires user.name and user.email to be set
  try {
    // If invocation was successful it means name exists
    try {
      Git(['config', '--global', 'user.name'])
      nameExists = true
    } catch (e) {
      Git(['config', '--global', 'user.name', 'tortilla'])
    }

    // If invocation was successful it means email exists
    try {
      Git(['config', '--global', 'user.email'])
      emailExists = true
    } catch (e) {
      Git(['config', '--global', 'user.email', 'tortilla@tortilla.academy'])
    }

    result = fn(...args)
  }
  finally {
    if (!nameExists) {
      Git(['config', '--global', '--unset', 'user.name'])
    }

    if (!emailExists) {
      Git(['config', '--global', '--unset', 'user.email'])
    }
  }

  return result
}

export const diffReleases = withUser((dump: object, srcRelease: string, destRelease: string) => {
  return Dump.diffReleases(dump, srcRelease, destRelease)
})

// In worker
if (!process.env.TORTILLA_WORKER) {
  // Listen for signals
  process.on('message', (msg: any) => {
    if (msg.op !== 'exec') { return }

    let handler

    switch (msg.fn) {
      case 'diffReleases': {
        handler = diffReleases; break
      }
      // Escape if function name is not supported
      default: return
    }

    try {
      const result = handler(...msg.args)

      process.send({
        op: 'result',
        fn: msg.fn,
        id: msg.id,
        payload: result,
      })
    }
    catch (e) {
      process.send({
        op: 'error',
        fn: msg.fn,
        id: msg.id,
        payload: e,
      })
    }
  })
}
