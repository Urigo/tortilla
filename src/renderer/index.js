const Fs = require('fs-extra');
const Handlebars = require('handlebars');
const Path = require('path');
const Git = require('../git');
const Paths = require('../paths');
const Release = require('../release');
const Utils = require('../utils');

/**
  A wrapper for Handlebars with several additions which are essential for Tortilla
 */

// Creating a new instance of handlebars which will then be merged with the module
const handlebars = Handlebars.create();
// Keep original handlers since these methods are gonna be overriden
const superRegisterHelper = handlebars.registerHelper.bind(handlebars);
const superRegisterPartial = handlebars.registerPartial.bind(handlebars);
// Used to store registered transformations
const transformations = {};
// Cache for templates which were already compiled
const cache = {};


// Read the provided file, render it, and overwrite it. Use with caution!
function overwriteTemplateFile(templatePath, scope) {
  templatePath = resolveTemplatePath(templatePath);
  const view = renderTemplateFile(templatePath, scope);

  return Fs.writeFileSync(templatePath, view);
}

// Read provided file and render its template. Note that the default path would
// be tortilla's template dir, so specifying a file name would be ok as well
function renderTemplateFile(templatePath, scope) {
  templatePath = resolveTemplatePath(templatePath);

  if (process.env.TORTILLA_CACHE_DISABLED || !cache[templatePath]) {
    const templateContent = Fs.readFileSync(templatePath, 'utf8');
    cache[templatePath] = handlebars.compile(templateContent);
  }

  const template = cache[templatePath];
  return renderTemplate(template, scope);
}

// Render provided template
function renderTemplate(template, scope) {
  // Template can either be a string or a compiled template object
  if (typeof template === 'string') {
    template = handlebars.compile(template);
  }
  scope = scope || {};

  if (scope.viewPath) {
    // Relative path of view dir
    // e.g. manuals/views
    var viewDir = Path.dirname(scope.viewPath);
  }

  const oldResolve = handlebars.resolve;

  try {
    // Set the view file for the resolve utility. If no view path was provided, the
    // resolve function below still won't work
    handlebars.resolve = resolvePath.bind(null, viewDir);
    return template(scope);
  }
  finally {
    // Either if an error was thrown or not, unbind it
    handlebars.resolve = oldResolve;
  }
}

// Returns a template path relative to tortilla with an '.tmpl' extension
function resolveTemplatePath(templatePath) {
  if (templatePath.indexOf('.tmpl') == -1) {
    templatePath += '.tmpl';
  }

  // User defined templates
  const relativeTemplatePath = Path.resolve(Paths.manuals.templates, templatePath);
  if (Utils.exists(relativeTemplatePath)) {
    return relativeTemplatePath;
  }

  // Tortilla defined templates
  return Path.resolve(Paths.tortilla.renderer.templates, templatePath);
}

// Register a new helper. Registered helpers will be wrapped with a
// [{]: <helper> (name ...args) [}]: #
function registerHelper(name, helper, options) {
  options = options || {};

  const wrappedHelper = function () {
    const oldCall = handlebars.call;

    try {
      // Bind the call method to the current context
      handlebars.call = callHelper.bind(this);
      var out = helper.apply(this, arguments);
    } finally { // Fallback
      // Restore method to its original
      handlebars.call = oldCall;
    }

    if (typeof out !== 'string' &&
        !(out instanceof String)) {
      throw Error([
        'Template helper', name, 'must return a string!',
        'Instead it returned', out,
      ].join(' '));
    }

    const target = process.env.TORTILLA_RENDER_TARGET;
    const args = [].slice.call(arguments);

    // Transform helper output
    const transformation = transformations[target] && transformations[target][name];
    if (transformation) {
      out = transformation(...[out].concat(args));
    }

    // Wrap helper output
    if (options.mdWrap) {
      out = mdWrapComponent('helper', name, args, out);
    }

    return out;
  };

  superRegisterHelper(name, wrappedHelper);
}

// Register a new partial. Registered partials will be wrapped with a
// [{]: <partial> (name) [}]: #
function registerPartial(name, partial, options) {
  options = options || {};

  // Wrap partial template
  if (options.mdWrap) {
    partial = mdWrapComponent('partial', name, partial);
  }

  return superRegisterPartial(name, partial);
}

// Register a new transformation which will take effect on rendered helpers. This is
// useful when setting the TORTILLA_RENDER_TARGET variable, so we can make additional
// adjustments for custom targets. For now this is NOT part of the official API and
// is used only for development purposes
function registerTransformation(targetName, helperName, transformation) {
  if (!transformations[targetName]) {
    transformations[targetName] = {};
  }
  transformations[targetName][helperName] = transformation;
}

// Returns content wrapped by component notations. Mostly useful if we want to detect
// components in the view later on using external softwares later on.
// e.g. https://github.com/Urigo/angular-meteor-docs/blob/master/src/app/tutorials/
// improve-code-resolver.ts#L24
function mdWrapComponent(type, name, args, content) {
  let hash = {};

  if (typeof content !== 'string') {
    content = args;
    args = [];
  }

  if (args[args.length - 1] instanceof Object) {
    hash = args.pop().hash;
  }

  // Stringify arguments
  const params = args.map(param => typeof param === 'string' ? `"${param}"` : param).join(' ');

  hash = stringifyHash(hash);

  // Concat all stringified arguments
  args = [name, params, hash]
    // Get rid of empty strings
    .filter(Boolean)
    .join(' ');

  return `[{]: <${type}> (${Utils.escapeBrackets(args)})\n\n${content}\n\n[}]: #`;
}

// Takes a helper hash and stringifying it
// e.g. { foo: '1', bar: 2 } -> foo="1" bar=2
function stringifyHash(hash) {
  return Object.keys(hash).map((key) => {
    let value = hash[key];
    if (typeof value === 'string') {
      value = `"${value}"`;
    }
    return `${key}=${value}`;
  }).join(' ');
}

// Calls a template helper with the provided context and arguments
function callHelper(methodName) {
  const args = [].slice.call(arguments, 1);
  let options = args.pop();

  // Simulate call from template
  if (options instanceof Object) {
    options = { hash: options };
  }

  args.push(options);

  return handlebars.helpers[methodName].apply(this, args);
}

// Takes a bunch of paths and resolved them relatively to the current rendered view
function resolvePath(/* reserved path, user defined path */) {
  let paths = [].slice.call(arguments);

  // A default path that the host's markdown renderer will know how to resolve by its own
  let defaultPath = paths.slice(1).join('/');
  defaultPath = new String(defaultPath);
  // The 'isRelative' flag can be used later on to determine if this is an absolute path
  // or a relative path
  defaultPath.isRelative = true;

  const cwd = paths.shift()

  // If function is unbound, return default path
  if (typeof cwd != 'string') {
    return defaultPath;
  }

  const repository = Fs.readJsonSync(Paths.npm.package).repository;

  // If no repository was defined, or
  // repository type is not git, or
  // no repository url is defined, return default path
  if (repository == null ||
      repository.type != 'git' ||
      repository.url == null) {
    return defaultPath;
  }

  const currentRelease = Release.format(Release.current());

  // Any release is yet to exist
  if (currentRelease == '0.0.0') {
    return defaultPath;
  }

  // Compose branch path for current release tree
  // e.g. github.com/Urigo/Ionic2CLI-Meteor-Whatsapp/tree/master@0.0.1
  const releaseTag = `${Git.activeBranchName()}@${currentRelease}`;
  const repositoryUrl = repository.url.replace('.git', '');
  const branchUrl = [repositoryUrl, 'tree', releaseTag].join('\/');
  const protocol = (branchUrl.match(/^.+\:\/\//) || [''])[0];
  const branchPath = `/${branchUrl.substr(protocol.length)}`;

  // If we use tilde (~) at the beginning of the path, we will be referenced to the
  // repo's root URL. This is useful when we want to compose links which are
  // completely disconnected from the current state, like commits, issues and PRs
  paths = paths
    .map(path => path.replace(/~/g, Path.resolve(branchPath, '../..')))
    .map(path => Path.isAbsolute(path) ? Path.relative(cwd, path) : Path.resolve(branchPath, Path.join(cwd, path)));

  // Resolve full path
  // e.g. github.com/Urigo/Ionic2CLI-Meteor-Whatsapp/tree/master@0.0.1
  // /manuals/views/step1.md
  paths.unshift(branchPath);
  return protocol + Path.resolve(...paths).substr(1);
}


module.exports = Utils.extend(handlebars, {
  overwriteTemplateFile,
  renderTemplateFile,
  renderTemplate,
  registerHelper,
  registerPartial,
  registerTransformation,
  // This should be set whenever we're in a helper scope
  call: callHelper,
  // Should be bound by the `renderTemplate` method
  resolve: resolvePath.bind(null, null),
});

// Built-in helpers and partials
require('./helpers/diff-step');
require('./helpers/nav-step');
require('./helpers/resolve-path');
require('./helpers/step-message');
require('./helpers/translate');
