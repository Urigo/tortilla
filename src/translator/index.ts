import * as i18n from 'i18next';
import * as Path from 'path';
import { Paths } from '../paths';
import { Utils } from '../utils';
import { Translation } from './translation';

const superTranslate = i18n.t.bind(i18n);

i18n.init({
  lng: 'en',
  initImmediate: true,
  resources: {
    en: { translation: getTranslationResource('en') },
    he: { translation: getTranslationResource('he') },
  },
});

// Gets a locale and returns the full resource object
function getTranslationResource(locale) {
  let paths: any = Paths;

  if (process.env.TORTILLA_CWD) {
    paths = Paths.resolveProject(process.env.TORTILLA_CWD);
  }

  paths = [
    // Static locales
    Path.resolve(paths.tortilla.translator.locales, `${locale}.json`),
    // User defined locales
    Path.resolve(paths.locales, `${locale}.json`),
  ];

  // Unite all resources and return a single one
  return paths.reduce((resource, path) => {
    // Clear cache so files can be properly reloaded
    try {
      delete require.cache[require.resolve(path)];
      const extension = require(path);
      Utils.merge(resource, extension);
    } catch (err) {
      // nothing to do here...
    }

    return resource;
  }, {});
}

// Returns i18n translation wrapped with some extra functionality
function translate(...args) {
  const result = superTranslate(...args);

  return result && new Translation(result);
}

// Any translation would be done using the provided locale
function scopeLanguage(language, fn) {
  if (!language) {
    return fn();
  }

  const oldLanguage = (i18n as any).translator.language;

  try {
    (i18n as any).translator.changeLanguage(language);
    fn();
  } finally {
    (i18n as any).translator.changeLanguage(oldLanguage);
  }
}

// Shallow cloning i18n so it won't be changed
export const Translator = Utils.extend(i18n, {
  translate,
  scopeLanguage,
});
