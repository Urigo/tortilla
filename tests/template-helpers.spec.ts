import { tortillaBeforeAll, tortillaBeforeEach } from './tests-helper';
import './custom-matchers';
import * as Fs from 'fs-extra';
import * as Path from 'path';
import { Renderer } from '../src/renderer';
import { Translator } from '../src/translator';

const Pack = require('../package.json');

let context: any = {};

describe('Template Helpers', () => {
  beforeAll(tortillaBeforeAll.bind(context));
  beforeEach(tortillaBeforeEach.bind(context));

  describe('diffStep', function() {
    it('should render an added file', function() {
      context.applyTestPatch('add-file');

      const view = Renderer.renderTemplate('{{{diffStep 1.1}}}');
      expect(view).toContainSameContentAsFile('add-file.md');
    });

    it('should escape file path with __pattern__', function() {
      context.applyTestPatch('add-private-file');

      const view = Renderer.renderTemplate('{{{diffStep 1.1}}}');
      expect(view).toContainSameContentAsFile('add-private-file.md');
    });

    it('should render changes made in file', function() {
      context.applyTestPatch('add-file');
      context.applyTestPatch('change-file');

      const view = Renderer.renderTemplate('{{{diffStep 1.2}}}');
      expect(view).toContainSameContentAsFile('change-file.md');
    });

    it('should render a deleted file', function() {
      context.applyTestPatch('add-file');
      context.applyTestPatch('delete-file');

      const view = Renderer.renderTemplate('{{{diffStep 1.2}}}');
      expect(view).toContainSameContentAsFile('delete-file.md');
    });

    it('should render a renamed file', function() {
      context.applyTestPatch('add-file');
      context.applyTestPatch('rename-file');

      const view = Renderer.renderTemplate('{{{diffStep 1.2}}}');
      expect(view).toContainSameContentAsFile('rename-file.md');
    });

    it('should render root diff', function() {
      const view = Renderer.renderTemplate('{{{diffStep "root"}}}');
      expect(view).toContainSameContentAsFile('root.md');
    });

    it('should render only the file matching the pattern', function() {
      context.applyTestPatch('add-multiple-files');

      const view = Renderer.renderTemplate('{{{diffStep 1.1 files="test-file.js"}}}');
      expect(view).toContainSameContentAsFile('add-multiple-files.md');
    });

    it('should render an error message if step not found', function() {
      context.applyTestPatch('add-file');

      const view = Renderer.renderTemplate('{{{diffStep 1.3}}}');
      expect(view).toContainSameContentAsFile('step-not-found.md');
    });

    it('should reference commit if repo URL is defined in package.json', function() {
      context.applyTestPatch('add-file');

      const packPath = context.exec('realpath', ['package.json']);
      const pack = Fs.readJsonSync(packPath);

      pack.repository = Pack.repository;
      Fs.writeFileSync(packPath, JSON.stringify(pack));

      const view = Renderer.renderTemplate('{{{diffStep 1.1}}}', {
        viewPath: 'dummy'
      });

      expect(view).toContainSameContentAsFile('referenced-diff.md');
    });

    it('should render step from the specified submodule', function() {
      const repoDir = context.createRepo();

      context.tortilla(['submodule', 'add', Path.basename(repoDir), repoDir]);

      const view = Renderer.renderTemplate(`{{{diffStep 1.1 module="${Path.basename(repoDir)}"}}}`);

      expect(view).toContainSameContentAsFile('submodule-diff.md');
    });

    it('should not render title if specified not to', function() {
      context.applyTestPatch('add-file');
      context.applyTestPatch('change-file');

      const view = Renderer.renderTemplate('{{{diffStep 1.2 noTitle=true}}}');
      expect(view).toContainSameContentAsFile('change-file-notitle.md');
    });

    describe('render target set to Medium', function() {
      beforeAll(function() {
        process.env.TORTILLA_RENDER_TARGET = 'medium';
      });

      afterAll(function() {
        delete process.env.TORTILLA_RENDER_TARGET;
      });

      it('should render an added file', function() {
        context.applyTestPatch('add-file');

        const view = Renderer.renderTemplate('{{{diffStep 1.1}}}');
        expect(view).toContainSameContentAsFile('medium/add-file.md');
      });

      it('should render changes made in file', function() {
        context.applyTestPatch('add-file');
        context.applyTestPatch('change-file');

        const view = Renderer.renderTemplate('{{{diffStep 1.2}}}');
        expect(view).toContainSameContentAsFile('medium/change-file.md');
      });

      it('should render only the file matching the pattern', function() {
        context.applyTestPatch('add-multiple-files');

        const view = Renderer.renderTemplate('{{{diffStep 1.1 files="test-file.js"}}}');
        expect(view).toContainSameContentAsFile('medium/add-multiple-files.md');
      });

      it('should render an error message if step not found', function() {
        context.applyTestPatch('add-file');

        const view = Renderer.renderTemplate('{{{diffStep 1.3}}}');
        expect(view).toContainSameContentAsFile('medium/step-not-found.md');
      });

      it('it should escape expressions', function() {
        context.applyTestPatch('change-view');

        const view = Renderer.renderTemplate('{{{diffStep 1.2}}}');
        expect(view).toContainSameContentAsFile('medium/escaped-diff.md');
      });
    });

    describe('translations', function() {
      beforeAll(function() {
        Translator.translator.changeLanguage('he');
      });

      beforeEach(function() {
        context.oldHEResource = Translator.getResourceBundle('he', 'translation');
      });

      afterEach(function() {
        Translator.addResourceBundle('he', 'translation', context.oldHEResource);
      });

      afterAll(function() {
        Translator.translator.changeLanguage('en');
      });

      it('should render an added file', function() {
        Translator.addResourceBundle('he', 'translation', {
          step: {
            1.1: 'הוספת קובץ'
          }
        });

        context.applyTestPatch('add-file');

        const view = Renderer.renderTemplate('{{{diffStep 1.1}}}');

        expect(view).toContainSameContentAsFile('he/add-file.md');
      });

      it('should render changes made in file', function() {
        Translator.addResourceBundle('he', 'translation', {
          step: {
            1.2: 'שינוי קובץ'
          }
        });

        context.applyTestPatch('add-file');
        context.applyTestPatch('change-file');

        const view = Renderer.renderTemplate('{{{diffStep 1.2}}}');
        expect(view).toContainSameContentAsFile('he/change-file.md');
      });

      it('should render a deleted file', function() {
        Translator.addResourceBundle('he', 'translation', {
          step: {
            1.2: 'מחיקת קובץ'
          }
        });

        context.applyTestPatch('add-file');
        context.applyTestPatch('delete-file');

        const view = Renderer.renderTemplate('{{{diffStep 1.2}}}');
        expect(view).toContainSameContentAsFile('he/delete-file.md');
      });

      it('should render a renamed file', function() {
        Translator.addResourceBundle('he', 'translation', {
          step: {
            1.2: 'שינוי שם קובץ'
          }
        });

        context.applyTestPatch('add-file');
        context.applyTestPatch('rename-file');

        const view = Renderer.renderTemplate('{{{diffStep 1.2}}}');
        expect(view).toContainSameContentAsFile('he/rename-file.md');
      });

      it('should render only the file matching the pattern', function() {
        Translator.addResourceBundle('he', 'translation', {
          step: {
            1.1: 'הוספת מיספר קבצים'
          }
        });

        context.applyTestPatch('add-multiple-files');

        const view = Renderer.renderTemplate('{{{diffStep 1.1 files="test-file.js"}}}');
        expect(view).toContainSameContentAsFile('he/add-multiple-files.md');
      });

      it('should render an error message if step not found', function() {
        context.applyTestPatch('add-file');

        const view = Renderer.renderTemplate('{{{diffStep 1.3}}}');
        expect(view).toContainSameContentAsFile('he/step-not-found.md');
      });
    });
  });

  describe('navStep', function() {
    beforeEach(function() {
      context.tortilla(['step', 'tag', '-m', 'dummy']);
      context.tortilla(['step', 'tag', '-m', 'dummy']);
      context.tortilla(['step', 'tag', '-m', 'dummy']);
    });

    it('should render a button referencing to the first step when editing the root commit', function() {
      context.tortilla(['step', 'edit', '--root']);

      const view = Renderer.renderTemplate('{{{navStep}}}');
      expect(view).toContainSameContentAsFile('begin-tutorial.md');
    });

    it('should render a button referencing to the second step when editing the first step', function() {
      context.tortilla(['step', 'edit', '1']);

      const view = Renderer.renderTemplate('{{{navStep}}}');
      expect(view).toContainSameContentAsFile('next-step.md');
    });

    it('should render a button referencing to the previous step when editing the last step', function() {
      context.tortilla(['step', 'edit', '3']);

      const view = Renderer.renderTemplate('{{{navStep}}}');
      expect(view).toContainSameContentAsFile('prev-step.md');
    });

    it('should render navigation buttons when editing a step in the middle of the stack', function() {
      context.tortilla(['step', 'edit', '2']);

      const view = Renderer.renderTemplate('{{{navStep}}}');
      expect(view).toContainSameContentAsFile('nav-steps.md');
    });

    it('should create navigation buttons with custom specified reference', function() {
      context.tortilla(['step', 'edit', '2']);

      const view = Renderer.renderTemplate('{{{navStep prevRef="http://test.com/prev/" nextRef="http://test.com/next/"}}}');
      expect(view).toContainSameContentAsFile('nav-steps-ref.md');
    });

    it('should create a single button referencing the README.md file', function() {
      context.tortilla(['step', 'pop']);
      context.tortilla(['step', 'pop']);

      const view = Renderer.renderTemplate('{{{navStep}}}');
      expect(view).toContainSameContentAsFile('prev-root.md');
    });

    describe('render target set to Medium', function() {
      beforeAll(function() {
        process.env.TORTILLA_RENDER_TARGET = 'medium';
      });

      afterAll(function() {
        delete process.env.TORTILLA_RENDER_TARGET;
      });

      it('should render a button referencing to the first step when editing the root commit', function() {
        context.tortilla(['step', 'edit', '--root']);

        const view = Renderer.renderTemplate('{{{navStep}}}');
        expect(view).toContainSameContentAsFile('medium/begin-tutorial.md');
      });

      it('should render a button referencing to the second step when editing the first step', function() {
        context.tortilla(['step', 'edit', '1']);

        const view = Renderer.renderTemplate('{{{navStep}}}');
        expect(view).toContainSameContentAsFile('medium/next-step.md');
      });

      it('should render a button referencing to the previous step when editing the last step', function() {
        context.tortilla(['step', 'edit', '3']);

        const view = Renderer.renderTemplate('{{{navStep}}}');
        expect(view).toContainSameContentAsFile('medium/prev-step.md');
      });

      it('should render navigation buttons when editing a step in the middle of the stack', function() {
        context.tortilla(['step', 'edit', '2']);

        const view = Renderer.renderTemplate('{{{navStep}}}');
        expect(view).toContainSameContentAsFile('medium/nav-steps.md');
      });

      it('should create navigation buttons with custom specified reference', function() {
        context.tortilla(['step', 'edit', '2']);

        const view = Renderer.renderTemplate('{{{navStep prevRef="http://test.com/prev/" nextRef="http://test.com/next/"}}}');
        expect(view).toContainSameContentAsFile('medium/nav-steps-ref.md');
      });

      it('should create a single button referencing the README.md file', function() {
        context.tortilla(['step', 'pop']);
        context.tortilla(['step', 'pop']);

        const view = Renderer.renderTemplate('{{{navStep}}}');
        expect(view).toContainSameContentAsFile('medium/prev-root.md');
      });
    });

    describe('translations', function() {
      beforeAll(function() {
        Translator.translator.changeLanguage('he');
      });

      beforeEach(function() {
        context.oldHEResource = Translator.getResourceBundle('he', 'translation');
      });

      afterEach(function() {
        Translator.addResourceBundle('he', 'translation', context.oldHEResource);
      });

      afterAll(function() {
        Translator.translator.changeLanguage('en');
      });

      it('should render a button referencing to the first step when editing the root commit', function() {
        context.tortilla(['step', 'edit', '--root']);

        const view = Renderer.renderTemplate('{{{navStep}}}');
        expect(view).toContainSameContentAsFile('he/begin-tutorial.md');
      });

      it('should render a button referencing to the second step when editing the first step', function() {
        context.tortilla(['step', 'edit', '1']);

        const view = Renderer.renderTemplate('{{{navStep}}}');
        expect(view).toContainSameContentAsFile('he/next-step.md');
      });

      it('should render a button referencing to the previous step when editing the last step', function() {
        context.tortilla(['step', 'edit', '3']);

        const view = Renderer.renderTemplate('{{{navStep}}}');
        expect(view).toContainSameContentAsFile('he/prev-step.md');
      });

      it('should render navigation buttons when editing a step in the middle of the stack', function() {
        context.tortilla(['step', 'edit', '2']);

        const view = Renderer.renderTemplate('{{{navStep}}}');
        expect(view).toContainSameContentAsFile('he/nav-steps.md');
      });

      it('should create navigation buttons with custom specified reference', function() {
        context.tortilla(['step', 'edit', '2']);

        const view = Renderer.renderTemplate('{{{navStep prevRef="http://test.com/prev/" nextRef="http://test.com/next/"}}}');
        expect(view).toContainSameContentAsFile('he/nav-steps-ref.md');
      });

      it('should create a single button referencing the README.md file', function() {
        context.tortilla(['step', 'pop']);
        context.tortilla(['step', 'pop']);

        const view = Renderer.renderTemplate('{{{navStep}}}');
        expect(view).toContainSameContentAsFile('he/prev-root.md');
      });
    });
  });
});
