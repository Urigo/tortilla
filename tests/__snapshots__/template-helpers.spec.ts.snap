// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Template Helpers table of contents renders correctly: render-toc 1`] = `
"- [Step 1: Create App](.tortilla/manuals/views/step1.md#step-1-create-app)
    - [Step 1.1: Create Meteor Project](.tortilla/manuals/views/step1.md#step-11-create-meteor-project)
    - [Step 1.2: Create Task Component](.tortilla/manuals/views/step1.md#step-12-create-task-component)
    - [Step 1.3: Create Sample Tasks](.tortilla/manuals/views/step1.md#step-13-create-sample-tasks)
    - [Step 1.4: Render Sample Tasks](.tortilla/manuals/views/step1.md#step-14-render-sample-tasks)
- [Step 2: Collections](.tortilla/manuals/views/step2.md#step-2-collections)
    - [Step 2.1: Create Tasks Collection](.tortilla/manuals/views/step2.md#step-21-create-tasks-collection)
    - [Step 2.2: Initialize Tasks Collection](.tortilla/manuals/views/step2.md#step-22-initialize-tasks-collection)
    - [Step 2.3: Render Tasks Collection](.tortilla/manuals/views/step2.md#step-23-render-tasks-collection)
- [Step 3: Forms and Events](.tortilla/manuals/views/step3.md#step-3-forms-and-events)
    - [Step 3.1: Create Task Form](.tortilla/manuals/views/step3.md#step-31-create-task-form)
    - [Step 3.2: Add Submit Handler](.tortilla/manuals/views/step3.md#step-32-add-submit-handler)
    - [Step 3.3: Show Newest Tasks First](.tortilla/manuals/views/step3.md#step-33-show-newest-tasks-first)
"
`;
